python/lsst/ts/wep/donutSizeCorrelator.py

```diff
@@ -0,0 +1,348 @@
+            y, x = np.where(binned == np.nanmax(binned))
+
+            # Undo binning and pad
+            x = binning_factor * x[0] + pad
+            y = binning_factor * y[0] + pad
+
+            # Crop around brightest pixel
+            H, W = image.shape
+            x0 = np.clip(x - length // 2, 0, W - length)
+            y0 = np.clip(y - length // 2, 0, H - length)
+            image = image[y0 : y0 + length, x0 : x0 + length]
+
+        # Bin array using median to avoid impact of hot pixels
+        if binning is not None:
+            image = binArray(image, binning, "median")
+
+        return image
+
+    @staticmethod
+    def createDonutTemplate(diameter: float) -> np.ndarray:
+        """Create simple annulus template for donuts.
+
+        Parameters
+        ----------
+        diameter : float
+            Diameter of donut.
+
+        Returns
+        -------
+        np.ndarray
+            Fractional donut mask
+        """
+        # Create grid of pixel centers
+        x = np.arange(diameter + 5, dtype=float)
+        x -= x.mean()
+        x, y = np.meshgrid(x, x)
+
+        # Build sub-pixel offsets
+        N_offsets = 10
+        offsets = np.arange(N_offsets + 1) / N_offsets - 0.5
+        dy, dx = np.meshgrid(offsets, offsets, indexing="ij")
+        dy = dy.reshape(-1, 1, 1)  # Reshape for broadcasting
+        dx = dx.reshape(-1, 1, 1)
+
+        # Distance to each subpixel
+        r = np.sqrt((x + dx) ** 2 + (y + dy) ** 2)
+
+        # Mask pixels by distance, then average number inside mask
+        inside = (r >= 0.61 * diameter / 2) & (r <= diameter / 2)
+        inside = inside.mean(axis=0).astype(float)
+
+        return inside
+
+    def correlateImage(
+        self,
+        image: np.ndarray,
+        resolution: int = 4,
+        d_min: int = 20,
+        d_max: int = 200,
+        length: int | None = None,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Correlate the image with donuts of different sizes.
+
+        Parameters
+        ----------
+        image : np.ndarray
+            Image containing donuts
+        resolution : int, optional
+            Resolution of donut diameter in pixels. Default is 4.
+        d_min : int, optional
+            Minimum donut diameter in pixels. Default is 20.
+        d_max : int, optional
+            Maximum donut diameter in pixels. Default is 200.
+        length : int or None, optional
+            Size of stamp to cutout from center of image.
+            Default is None.
+
+        Returns
+        -------
+        np.ndarray
+            Array of diameters in pixels
+        np.ndarray
+            Array of correlation values
+        """
+        # Crop and bin the image
+        image = self.cropAndBinImage(
+            image=image,
+            length=length,
+            pad=min(d_max // 2, np.min(image.shape) // 4),
+            binning=resolution,
+        )
+
+        # Start with minimum diameter
+        diam = d_min
```

Comment:
This line looks extraneous? you are overwriting the variable in the for loop later